contracts/sophia.md

```diff
@@ -122,6 +122,82 @@ function root(t : tree('a)) : option('a) =
     Bin(_, v, _) => Some(v)
 ```
 
+### Maps and records
+
+A Sophia record type is given by a fixed set of fields with associated,
+possibly different, types. For instance
+```
+  type account = { name    : string,
+                   balance : int,
+                   history : list(transaction) }
+```
+
+Maps, on the other hand, can contain an arbitrary number of key-value bindings,
+but of a fixed type. The type of maps with keys of type `'k` and values of type
+`'v` is written `map('k, 'v)`. Key types are restricted to atomic types (`int`,
+`address`, `bool`, and `string`).
+
+#### Constructing maps and records
+
+A value of record type is constructed by giving a value for each of the fields.
+For the example above,
+```
+  function new_account(name) =
+    {name = name, balance = 0, history = []}
+```
+Maps are constructed similarly, with keys enclosed in square brackets
+```
+  function example_map() : map(string, int) =
+    {["key1"] = 1, ["key2"] = 2}
+```
+The empty map is written `{}`.
+
+#### Accessing values
+
+Record fields access is written `r.f` and map lookup `m[k]`. For instance,
+```
+  function get_balance(a : address, accounts : map(address, account)) =
+    accounts[a].balance
+```
+Looking up a non-existing key in a map results in the contract failing. See
+`Map.member` and `Map.lookup` below for safe lookups.
+
+#### Updating a value
+
+Record field updates are written `r{f = v}`. This creates a new record value
+which is the same as `r`, but with the value of the field `f` replaced by `v`.
+Similarly, `m{[k] = v}` constructs a map with the same values as `m` except
+that `k` maps to `v`. It makes no difference if `m` has a mapping for `k` or
```

Comment:
So if user would like to avoid new key by mistake/typo, he/she shall define auxillary function like "map update" asserting that key is member before updating value. Ok.

Reply:
Yes, or use the `@`-construct, which does require the key to be present: `m{ [k] @ _ = v }`.